File: presenie/src/main/scala/inosion/presenie/Presenie.scala
package inosion.presenie

import java.io.File
import org.rogach.scallop._


object Presenie extends App {

  val version = "1.0.0"

  class Conf(arguments: Seq[String]) extends ScallopConf(arguments) {


      trait TemplateArg { _: ScallopConf  =>
        val template = opt[File](required = true, descr = "The template to merge data with")
        mainOptions = Seq(template)
      }
      version(s"Presenie ${Presenie.version} (c) 2024 Inosion")
      val verbose =  opt[Boolean]()

      val list =     new Subcommand("list") with TemplateArg {}
      addSubcommand(list)

      val merge =     new Subcommand("merge") with TemplateArg {
        val data =     opt[File](required = true, descr = "The data file to merge with the template")
        val config =   opt[File](required = false, descr = "Config file to drive merging")
        val outFile =  opt[File](required = true, descr = "Output filename")
      }
      addSubcommand(merge)

      val cloner =     new Subcommand("clone") with TemplateArg {
        val outFile =  opt[File](required = true, descr = "Output filename")
      }
      addSubcommand(cloner)

      verify()

  }

  val conf = new Conf(args)

  conf.subcommand match {
    case Some(conf.list) => pptx.PPTXTools.listSlideLayouts(conf.list.template())
    case Some(conf.merge) => {
      pptx.PPTXMerger.render(conf.merge.data(), conf.merge.template(), conf.merge.outFile())
      Console.println(s"Wrote ${conf.merge.outFile.apply().getAbsolutePath()}")
    }
    case Some(conf.cloner) => {
      Console.println(s"Cloned ${conf.merge.template.apply().getAbsolutePath()} --> ${conf.merge.outFile.apply().getAbsolutePath()}")
      pptx.PPTXTools.clonePptx(conf.cloner.template().toPath(), conf.cloner.outFile().toPath())
    }
    case _ => conf.printHelp()
  }

}

File: src/presenie.py
import argparse
import sys
from pptx_tools import PPTXTools
from pptx_merger import PPTXMerger

VERSION = "1.0.0"

def main():
    parser = argparse.ArgumentParser(description=f"Presenie {VERSION} (c) 2024 Inosion")
    parser.add_argument('--verbose', action='store_true', help='Enable verbose output')

    subparsers = parser.add_subparsers(dest='command')

    # List subcommand
    list_parser = subparsers.add_parser('list', help='List slide layouts')
    list_parser.add_argument('--template', type=str, required=True, help='The template to merge data with')

    # Merge subcommand
    merge_parser = subparsers.add_parser('merge', help='Merge data with template')
    merge_parser.add_argument('--template', type=str, required=True, help='The template to merge data with')
    merge_parser.add_argument('--data', type=str, required=True, help='The data file to merge with the template')
    merge_parser.add_argument('--config', type=str, help='Config file to drive merging')
    merge_parser.add_argument('--outFile', type=str, required=True, help='Output filename')

    # Clone subcommand
    clone_parser = subparsers.add_parser('clone', help='Clone PPT slides')
    clone_parser.add_argument('--template', type=str, required=True, help='The template to merge data with')
    clone_parser.add_argument('--outFile', type=str, required=True, help='Output filename')

    args = parser.parse_args()

    if args.command == 'list':
        PPTXTools.list_slide_layouts(args.template)

    elif args.command == 'merge':
        PPTXMerger.render(args.data, args.template, args.outFile)
        print(f"Wrote {args.outFile}")

    elif args.command == 'clone':
        print(f"Cloned {args.template} --> {args.outFile}")
        PPTXTools.clone_pptx(args.template, args.outFile)
        
    else:
        parser.print_help()

if __name__ == "__main__":
    main()